src/Test/HUnit/Lang.hs

```diff
@@ -56,7 +62,21 @@ assertFailure ::
      HasCallStack =>
      String -- ^ A message that is displayed with the assertion failure
   -> Assertion
-assertFailure msg = msg `deepseq` E.throwIO (HUnitFailure location $ Reason msg)
+assertFailure = assertFailure'
+
+-- | Generalized version of `assertFailure`. Useful for signaling a failure in
+-- a code that expects a non-unit value. E.g.
+--
+-- @
+--    x <- if conditionIsMet
+--           then return e
+--           else assertFailure' msg
+-- @
+assertFailure' ::
+     HasCallStack =>
+     String -- ^ A message that is displayed with the assertion failure
+  -> Assertion' a
+assertFailure' msg = msg `deepseq` E.throwIO (HUnitFailure location $ Reason msg)
```

Comment:
Do you want me to change the type `Assertion` too or do you want to keep the the old `Assertion` and make `assertFailure` `Assertion' a` ?